aneekee/nodejs-aws-shop-backend
# task-6

Frontend -- https://d1vdv7tmd1pu8.cloudfront.net/admin/products

## How to test

1. Create a csv file, for example
```csv
title,description,price,count
JoJo's Bizarre Adventure,Adventure manga by Hirohiko Araki,21.99,13
```

2. Open the https://d1vdv7tmd1pu8.cloudfront.net/admin/products and upload your file. Please note that `catalogBatchProcess` will not write to the database if your csv file has a different structure.
3. Open the https://d1vdv7tmd1pu8.cloudfront.net and check the products list -- it should contain your items. Because of the [maxBatchingWindow](https://github.com/slnchn/nodejs-aws-shop-backend/pull/4/files#diff-6bdd62ea8d555605583305460a6b7c64658c3df4443ca9a57ce039bb4e9a065aR175) it'll not happen immediately, so please wait for about a minute
4. If you want me to send you screenshot of my mailbox, please let me know :)


## [Tasks](https://github.com/rolling-scopes-school/aws/blob/main/aws-developer/06_async_microservices_communication/task.md#tasks)

### Task 6.1

- [x] 1. Create a lambda function called catalogBatchProcess under the Product Service which will be triggered by an SQS event.
Done -- https://github.com/slnchn/nodejs-aws-shop-backend/pull/4/files#diff-6bdd62ea8d555605583305460a6b7c64658c3df4443ca9a57ce039bb4e9a065aR156-R177
- [x] 2. Create an SQS queue called catalogItemsQueue, in the AWS CDK Stack.
Done -- https://github.com/slnchn/nodejs-aws-shop-backend/pull/4/files#diff-6bdd62ea8d555605583305460a6b7c64658c3df4443ca9a57ce039bb4e9a065aR118-R124
- [x] 3. Configure the SQS to trigger lambda catalogBatchProcess with 5 messages at once via batchSize property.
Done -- https://github.com/slnchn/nodejs-aws-shop-backend/pull/4/files#diff-6bdd62ea8d555605583305460a6b7c64658c3df4443ca9a57ce039bb4e9a065aR174
- [x] 4. The lambda function should iterate over all SQS messages and create corresponding products in the products table.
Done -- https://github.com/slnchn/nodejs-aws-shop-backend/pull/4/files#diff-cc0a566c26f46dc974b0b00755d8b67124a0e58455c10e67e04170c73cf32448R82-R92

### Task 6.2

- [x] 1. Update the importFileParser lambda function in the Import Service to send each CSV record into SQS.
Done -- the [call](https://github.com/slnchn/nodejs-aws-shop-backend/pull/4/files#diff-f65f3263ec3455f62ba02c3424cd5c90271e5a3c5dfc448ca657c6f762f8741dR20-R21), and the [implementation](https://github.com/slnchn/nodejs-aws-shop-backend/pull/4/files#diff-9ee9500d665d3520a75f676b6f438ac82232ff7f3365283e8a93aac5f9c0174cR76-R121)
- [x] 2. It should no longer log entries from the readable stream to CloudWatch.
Done -- it doesn't call the [logFileStream](https://github.com/slnchn/nodejs-aws-shop-backend/pull/4/files#diff-f65f3263ec3455f62ba02c3424cd5c90271e5a3c5dfc448ca657c6f762f8741dL19) anymore :)

### Task 6.3

- [x] 1. Create an SNS topic createProductTopic and email subscription in the AWS CDK Stack of the Product Service.
Done -- the [topic](https://github.com/slnchn/nodejs-aws-shop-backend/pull/4/files#diff-6bdd62ea8d555605583305460a6b7c64658c3df4443ca9a57ce039bb4e9a065aR126-R132), and [subscriptions](https://github.com/slnchn/nodejs-aws-shop-backend/pull/4/files#diff-6bdd62ea8d555605583305460a6b7c64658c3df4443ca9a57ce039bb4e9a065aR136-R154)
- [x] 2. Create a subscription for this SNS topic with an email endpoint type with your own email in there.
Done -- https://github.com/slnchn/nodejs-aws-shop-backend/pull/4/files#diff-6bdd62ea8d555605583305460a6b7c64658c3df4443ca9a57ce039bb4e9a065aR137
- [x] 3. Update the catalogBatchProcess lambda function in the Product Service to send an event to the SNS topic once it creates products.
Done -- the [call](https://github.com/slnchn/nodejs-aws-shop-backend/pull/4/files#diff-cc0a566c26f46dc974b0b00755d8b67124a0e58455c10e67e04170c73cf32448R91), and the [implementation](https://github.com/slnchn/nodejs-aws-shop-backend/pull/4/files#diff-d55af6f5f20781889f45a021cefd369f51aca045793c5f9977f1a8f3f15e4d12R20-R43)

### Task 6.4

- [x] 1. Commit all your work to separate branch (e.g. task-6 from the latest master) in your own repository.
- [x] 2. Create a pull request to the master branch.
- [x] 3. Submit link to the pull request to Crosscheck page in [RS App](https://app.rs.school/).


### Additional tasks

- [x] catalogBatchProcess lambda is covered by unit tests
Done -- https://github.com/slnchn/nodejs-aws-shop-backend/pull/4/files#diff-b3a0bc158f82441a03febac3e0a862a47fa2470823bb1bc41d1ea53ce21ac402
- [x] set a Filter Policy for SNS createProductTopic in AWS CDK Stack and create an additional email subscription to distribute messages to different emails depending on the filter for any product attribute
Done -- https://github.com/slnchn/nodejs-aws-shop-backend/pull/4/files#diff-6bdd62ea8d555605583305460a6b7c64658c3df4443ca9a57ce039bb4e9a065aR136-R154
